### PR TITLE
add build caching option

### PR DESCRIPTION
**Pull request by DavidKorczynski**
First touch on https://github.com/google/oss-fuzz-gen/issues/499

Depends on: https://github.com/google/oss-fuzz/pull/12284

The way this work is by saving a cached version of `build_fuzzers` post running of `compile` and then modifying the Dockerfiles of a project to use this cached build image + an adjusted build script.

For example, for brotli the Dockerfile is originally:

```sh
                                                                                
FROM gcr.io/oss-fuzz-base/base-builder                                          
RUN apt-get update && apt-get install -y cmake libtool make                     
                                                                                
RUN git clone --depth 1 https://github.com/google/brotli.git                    
WORKDIR brotli                                                                  
COPY build.sh $SRC/                                                             
                                                                                
COPY 01.c /src/brotli/c/fuzz/decode_fuzzer.c      
```

a Dockerfile is then created which relies on the cached version, and it loosk like:

```sh
FROM cached_image_brotli                                                        
# RUN apt-get update && apt-get install -y cmake libtool make                   
#                                                                               
# RUN git clone --depth 1 https://github.com/google/brotli.git                  
# WORKDIR brotli                                                                
# COPY build.sh $SRC/                                                           
#                                                                               
COPY 01.c /src/brotli/c/fuzz/decode_fuzzer.c                                    
#                                                                               
COPY adjusted_build.sh $SRC/build.sh 
```

`adjusted_build.sh` is then the script that only builds fuzzers. This means we can also use `build_fuzzers`/`compile` workflows as we know it.

More specifically, this PR:

- Makes it possible to build Docker images of fuzzer build containers. Does this by running `build_fuzzers`, saving the docker container and then commit the docker container to an image. This image will have a projects' build set up post running of `compile`. This is then used when building fuzzers by OFG.
- Supports only ASAN mode for now. Should be easy to extend to coverage too.
- Currently builds images first and then uses them locally. We could extend, probably on another step of this, to use containers pushed by OSS-Fuzz itself.
- Only does the caching if a "cache-build-script" exists (added a few for some projects) which contains the build instructions post-build process. It should be easy to extend such that we can rely on some DB of auto-generated build scripts as well (ref: https://github.com/google/oss-fuzz/pull/11937) but I think it's nice to have both the option of us creating the scripts ourselves + an auto-generated DB.

### CODE REVIEW

**Comment by DavidKorczynski** [comment]
Can confirm from local runs that this saves time. I assume a lot of time will be saved by (1) using OSS-Fuzz bots to generated the containers; (2) extend to coverage as well.

File: run_all_experiments.py
Comment (DonggeLiu) [inline run_all_experiments.py:150]: I reckon we can build the cache image of all projects every day (if we have their `fuzzer_build_script` in your solution, or address the TODOs in mine).
OSS-Fuzz is already building all original projects anyway.

We run multiple experiments every day, no need to rebuild the cache images in every experiment.

**Comment by DavidKorczynski** [comment]
> Would you plan to support more projects before merging?

Yep. I'm also planning on supporting caching of code coverage before merging. I was asking for review from a perspective of the architecture. In short, I agree on https://github.com/google/oss-fuzz-gen/pull/514#discussion_r1696736532 and https://github.com/google/oss-fuzz-gen/pull/514#discussion_r1696731446 but thought this would be a good start

File: fuzzer_build_script/brotli
Comment (DavidKorczynski) [reply fuzzer_build_script/brotli:4]: It's actually fine that `make` is used because in most, if not all (am not sure if time sometimes matters), times `make` will only recompile the targets that have change, i.e. the harness

File: experiment/builder_runner.py
Comment (DavidKorczynski) [reply experiment/builder_runner.py]: Done, i've now renamed caches so the images are called e.g.:

```
gcr.io/oss-fuzz/libraw_coverage_cache
gcr.io/oss-fuzz/libraw_coverage_cache
```

Note that there is difference between container/image names, and container names have some additional naming limitations. However, we the containers are still temporary and it's the image names we care about.

File: experiment/builder_runner.py
Comment (DavidKorczynski) [reply experiment/builder_runner.py]: done

File: experiment/builder_runner.py
Comment (DavidKorczynski) [reply experiment/builder_runner.py:574]: done

File: run_all_experiments.py
Comment (DavidKorczynski) [reply run_all_experiments.py]: I'm fine with that, I say we land it now as is and then do the migration of OSS-Fuzz build images/images build from cloud in the next steps. It's working well locally and I think that's a nice stage

File: run_all_experiments.py
Comment (DavidKorczynski) [reply run_all_experiments.py]: done, thanks

File: experiment/builder_runner.py
Comment (DavidKorczynski) [reply experiment/builder_runner.py]: done, thanks

File: experiment/builder_runner.py
Comment (DavidKorczynski) [reply experiment/builder_runner.py]: done, thanks

File: experiment/builder_runner.py
Comment (DavidKorczynski) [reply experiment/builder_runner.py]: done, thanks

**Comment by DavidKorczynski** [comment]
> > Would you plan to support more projects before merging?
> 
> Yep. I'm also planning on supporting caching of code coverage before merging. I was asking for review from a perspective of the architecture. In short, I agree on [#514 (comment)](https://github.com/google/oss-fuzz-gen/pull/514#discussion_r1696736532) and [#514 (comment)](https://github.com/google/oss-fuzz-gen/pull/514#discussion_r1696731446) but thought this would be a good start

@DonggeLiu this has now been achieved. I think the PR is now in a good state to land in that it works well on local runs.

next steps will be:
- Tie this together with cached builds uploaded by OSS-Fuzz
- Migrate changes onto the cloud environment

I would prefer to do both of these steps in follow-ups

**Comment by DavidKorczynski** [comment]
/gcbrun exp -n dk-5439 -m vertex_ai_gemini-1-5 -b comparison

**Comment by DonggeLiu** [comment]
Cached image generation is almost ready on the OSS-Fuzz side: https://github.com/google/oss-fuzz/pull/11937

Once merged, we can integrate it into the automatic daily build workflow.

File: run_all_experiments.py
Comment (DonggeLiu) [inline run_all_experiments.py:153]: nit: `for sanitizer in ['address', 'coverage']:`.

File: fuzzer_build_script/bluez
Comment (DonggeLiu) [inline fuzzer_build_script/bluez:1]: In a later PR, would it be better to store this in the benchmark YAML file?

**Review by DonggeLiu — APPROVED** [review]
Thanks @DavidKorczynski !

File: fuzzer_build_script/brotli
Comment (oliverchang) [reply fuzzer_build_script/brotli:4]: +1, let's keep this as more of an "override" option that's used in if the automatic solution fails.

File: fuzzer_build_script/bluez
Comment (oliverchang) [reply fuzzer_build_script/bluez:1]: Or better, can we store these upstream in oss-fuzz instead with some kind of convention? Are these reusable enough such that the upstream `build.sh` can call these so there's a single implementation of build script everywhere? 

Storing them here will inevitably cause things to diverge over time.

Comment (DonggeLiu) [reply fuzzer_build_script/bluez:1]: Oh nice, OSS-Fuzz is definitely a better place.
If the upstream project maintainers do not mind, maybe we can separate this part from `build.sh` into a new file:
```bash
# Original commands to build the project
...

# Call this script to generate fuzz target binaries.
/src/build_fuzzer
```

File: experiment/builder_runner.py
Comment (DavidKorczynski) [reply experiment/builder_runner.py:588]: We don't need `COPY` for this, as the dictionaries will already existing due to the containers being a snapshot *post* the first running of the original Dockerfile.

File: fuzzer_build_script/bluez
Comment (DavidKorczynski) [reply fuzzer_build_script/bluez:1]: I probably prefer keeping as is rather than storing in benchmark or oss-fuzz.

I don't think it will be a huge burden to maintain these, and if the auto-gen works well then we can most likely get rid of the many manual build scripts.

I'm not too much for putting it in OSS-Fuzz as I don't think it will necessarily be intuitive why it's there, and I would probably think it's best not to increase complexity of the fuzzing build setups in OSS-Fuzz. it's kind of mixing things a bit IMO.

I lean towards storing them here for now -- and if after some time it's clear storing them in OSS-Fuzz provides much benefit then we can move them there.

File: experiment/builder_runner.py
Comment (DavidKorczynski) [reply experiment/builder_runner.py]: yep, fixed the naming.

File: run_all_experiments.py
Comment (DavidKorczynski) [reply run_all_experiments.py]: moved them to oss_fuzz_checkout

**Comment by DavidKorczynski** [comment]
/gcbrun exp -n dk-5440 -m vertex_ai_gemini-1-5 -b comparison

File: experiment/builder_runner.py
Comment (DonggeLiu) [reply experiment/builder_runner.py]: Thanks for the explanation about no need to `COPY` dictionary/options.
My primary concern is the first sentence: I meant to say 'The code above does not reliably store the lines that `COPY`s fuzzer and build script'.

IIUC, the code assumes `copy_build_line` is always the last line starting with `COPY`, and the `copy_fuzzer_line` is always the second last. The links I added are some counter-examples to show this assumption is unreliable. For example, in the first case, the second last `COPY` is `COPY sksl.dict $SRC/skia_data/sksl.dict`.

If we use this on more projects later, then projects will be tripped up by this.

Comment (DavidKorczynski) [reply experiment/builder_runner.py]: Ah no, what these two lines are *not* commenting out is:

- the `COPY` added by OFG which copies in the new auto-generated harness
- the `COPY` added for the adjusted build script `COPY adjusted_build.sh $SRC/build.sh`

It's always guaranteed to be those two

File: experiment/builder_runner.py
Comment (DonggeLiu) [reply experiment/builder_runner.py:550]: Not yet moved, I presume?
I guess the simplest way is to encapsulate this into a function, host the function in `oss_fuzz_checkout`, and call it here.

File: run_all_experiments.py
Comment (DonggeLiu) [reply run_all_experiments.py]: Do you plan to make them private functions?

Comment (DavidKorczynski) [reply run_all_experiments.py]: I'm indifferent -- I didn't at first because they're no longer functions as part of a class.

Let me know which you prefer and I'll do that.

Comment (DonggeLiu) [reply run_all_experiments.py]: Thanks, I prefer making them private. Then if some of them are not called anymore in the future, our linter can know and remind us to remove them.

**Comment by DavidKorczynski** [comment]
/gcbrun exp -n dk-5442 -m vertex_ai_gemini-1-5 -b comparison

File: run_all_experiments.py
Comment (DavidKorczynski) [reply run_all_experiments.py]: Sounds good. Done

File: fuzzer_build_script/bluez
Comment (DonggeLiu) [reply fuzzer_build_script/bluez:1]: For OFG's need, I am not very confident that [chronos](https://github.com/google/oss-fuzz/pull/11937) will work for all projects immediately, so we might need manual scripts for a while.
A benefit of storing them in OSS-Fuzz is that Cloud Build instances can directly use them from OSS-Fuzz without needing to download them from OFG.

@oliverchang Is there any other way that OSS-Fuzz can benefit from these files?

Comment (DavidKorczynski) [reply fuzzer_build_script/bluez:1]: Either case should work, but I still think it would be, at least for now, good to not land them in OSS-Fuzz projects' folders until we have a stable set up.

I'll look to land this as it's useful as is and then we can adjust the OSS-Fuzz set up in further steps.

**Pull request merged**
